In [1]:
import numpy as np
from sklearn.linear_model import Perceptron

In [2]:
data = np.array([
    [1,1,0,1,0],
    [1,1,1,0,1],
    [0,0,1,0,0],
    [0,1,0,1,0],
    [1,0,1,1,1],
    [0,1,1,1,0]
])

In [3]:
data.shape

(6, 5)

In [4]:
x = data[:,:-1]
y = data[:,-1]

In [16]:
eta = (0.01, 0.1, 1.0)
for e in eta:
    print(f"Learning rate: {e}")
    for i in range(1,6):
        model = Perceptron(
            max_iter= i, 
            tol=1e-3, 
            eta0=e,
            random_state=42, 
            # verbose=True, 
            early_stopping=False,)
        
        model.fit(x,y)
        print(f"  -epoch: {i}, Score: {model.score(x,y)}")



Learning rate: 0.01
  -epoch: 1, Score: 0.8333333333333334
  -epoch: 2, Score: 0.6666666666666666
  -epoch: 3, Score: 0.8333333333333334
  -epoch: 4, Score: 0.8333333333333334
  -epoch: 5, Score: 0.8333333333333334
Learning rate: 0.1
  -epoch: 1, Score: 0.8333333333333334
  -epoch: 2, Score: 0.6666666666666666
  -epoch: 3, Score: 0.8333333333333334
  -epoch: 4, Score: 1.0
  -epoch: 5, Score: 0.8333333333333334
Learning rate: 1.0
  -epoch: 1, Score: 0.8333333333333334
  -epoch: 2, Score: 0.6666666666666666
  -epoch: 3, Score: 1.0
  -epoch: 4, Score: 1.0
  -epoch: 5, Score: 0.8333333333333334


/Users/qiyunge/winMac/SoftwareDev/LaselleCollege/3_C2_gestion_de_données_avancé/.venv/lib/python3.12/site-packages/sklearn/linear_model/_stochastic_gradient.py:741: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(
/Users/qiyunge/winMac/SoftwareDev/LaselleCollege/3_C2_gestion_de_données_avancé/.venv/lib/python3.12/site-packages/sklearn/linear_model/_stochastic_gradient.py:741: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(
/Users/qiyunge/winMac/SoftwareDev/LaselleCollege/3_C2_gestion_de_données_avancé/.venv/lib/python3.12/site-packages/sklearn/linear_model/_stochastic_gradient.py:741: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(
/Users/qiyunge/winMac/SoftwareDev/LaselleCollege/3_C2_gestion_de_données_avancé

Lorsque le nombre d’époques augmente de 1 à 5, le modèle a plus d’occasions de mettre à jour ses poids.

Avec eta = 0.01, le score reste généralement à 0.8333, sauf à l’époque 2 où il descend à 0.6667. Cela montre que le modèle n’est pas encore complètement convergé.

Avec eta = 0.1, le score atteint 1.0 à l’époque 4, mais redescend ensuite à 0.8333 à l’époque 5. Cela indique une convergence temporaire, mais pas totalement stable.

Avec eta = 1.0, le modèle atteint 1.0 aux époques 3 et 4, puis redescend aussi à 0.8333. Le taux d’apprentissage plus élevé permet donc d’atteindre plus rapidement un bon score, mais le résultat reste instable.

Les avertissements de convergence indiquent que le nombre maximal d’itérations est atteint avant une convergence complète. Donc, entre 1 et 5 époques, le modèle commence à apprendre, mais il ne converge pas encore de manière stable.